libs/wire-api/src/Wire/API/User/Auth.hs

```diff
@@ -40,102 +43,85 @@ module Wire.API.User.Auth
+  | LoginByHandle Handle
   deriving stock (Eq, Show, Generic)
-  deriving (Arbitrary) via (GenericUniform Login)
-
-modelLogin :: Doc.Model
-modelLogin = Doc.defineModel "Login" $ do
-  Doc.description "Payload for performing a login."
-  Doc.property "email" Doc.string' $ do
-    Doc.description "The email address for a password login."
-    Doc.optional
-  Doc.property "phone" Doc.string' $ do
-    Doc.description "The phone number for a password or SMS login."
-    Doc.optional
-  Doc.property "handle" Doc.string' $ do
-    Doc.description "The handle for a password login."
-    Doc.optional
-  Doc.property "password" Doc.string' $ do
-    Doc.description "The password for a password login."
-    Doc.optional
-  Doc.property "code" Doc.string' $ do
-    Doc.description "The login code for an SMS login."
-    Doc.optional
-  Doc.property "label" Doc.string' $ do
-    Doc.description
-      "A label to associate with the returned cookie. \
-      \Every client should have a unique and stable (persistent) label \
-      \to allow targeted revocation of all cookies granted to that \
-      \specific client."
-    Doc.optional
-  Doc.property "verification_code" Doc.string' $ do
-    Doc.description "The login verification code for 2nd factor authentication. Required only if SndFactorPasswordChallenge is enabled for the team/server."
-    Doc.optional
-
-instance ToJSON Login where
-  toJSON (SmsLogin p c l) = object ["phone" .= p, "code" .= c, "label" .= l]
-  toJSON (PasswordLogin login password label mbCode) =
-    object
-      [ "password" .= password,
-        "label" .= label,
-        loginIdPair login,
-        "verification_code" .= mbCode
-      ]
-
-instance FromJSON Login where
-  parseJSON = withObject "Login" $ \o -> do
-    passw <- o .:? "password"
-    case passw of
-      Nothing ->
-        SmsLogin <$> o .: "phone" <*> o .: "code" <*> o .:? "label"
-      Just pw -> do
-        loginId <- parseJSON (Object o)
-        PasswordLogin loginId pw <$> (o .:? "label") <*> (o .:? "verification_code")
-
-loginLabel :: Login -> Maybe CookieLabel
-loginLabel (PasswordLogin _ _ l _) = l
-loginLabel (SmsLogin _ _ l) = l
-
-loginIdPair :: LoginId -> Aeson.Pair
-loginIdPair = \case
-  LoginByEmail s -> "email" .= s
-  LoginByPhone s -> "phone" .= s
-  LoginByHandle s -> "handle" .= s
+  deriving (Arbitrary) via (GenericUniform LoginId)
+  deriving (FromJSON, ToJSON, S.ToSchema) via Schema LoginId
+
+-- NB. this should fail if (e.g.) the email is present but unparseable even if the JSON contains a valid phone number or handle.
+-- See tests in `Test.Wire.API.User.Auth`.
+instance ToSchema LoginId where
+  schema = object "LoginId" $ loginObjectSchema
+
+loginObjectSchema :: ObjectSchema SwaggerDoc LoginId
+loginObjectSchema =
+  fromLoginId .= tupleSchema `withParser` validate
+  where
+    fromLoginId :: LoginId -> (Maybe Email, Maybe Phone, Maybe Handle)
+    fromLoginId = \case
+      LoginByEmail e -> (Just e, Nothing, Nothing)
```

Comment:
There is a better way to do this with schema-profunctor, but it has a slightly different semantics in case of errors. With the current approach, an invalid email causes the whole thing to fail, even if a phone is present. If we use the standard schema-profunctor approach, based on `tag`, then an invalid email is treated the same as a missing email.

This is a general problem with parser combinators, since they can make it rather tricky to specify when to "commit" to a choice. We have discussed this sort of thing a lot in the past, and I don't think it has any good solution, as long as we're using Aeson under the hood. Ideally, it would be nice to completely remove Aeson from the equation, and use a better-designed parser that is able to distinguish between "fatal" and "non-fatal" errors. As it is, we're stuck with playing these sorts of games.